osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs

```diff
@@ -32,20 +38,30 @@ protected sealed override void UpdateState(ArmedState state)
             {
                 UpdatePreemptState();
 
-                using (BeginDelayedSequence(TIME_PREEMPT + (Judgements.FirstOrDefault()?.TimeOffset ?? 0), true))
+                var delay = TIME_PREEMPT + (Judgements.FirstOrDefault()?.TimeOffset ?? 0) - EarlyFadeOutTime;
+                using (BeginDelayedSequence(delay, true))
+                {
                     UpdateCurrentState(state);
+                    UpdatePostState();
+                }
             }
         }
 
         protected virtual void UpdatePreemptState()
         {
-            this.FadeIn(TIME_FADEIN);
+            this.FadeIn(TIME_FADEIN / FadeInSpeed);
         }
 
         protected virtual void UpdateCurrentState(ArmedState state)
         {
         }
 
+        protected virtual void UpdatePostState()
+        {
+            double duration = ((HitObject as IHasEndTime)?.EndTime ?? HitObject.StartTime) - HitObject.StartTime + EarlyFadeOutTime;
+            this.Delay(duration).Expire();

```

Comment:
Isn't this going to break for non-hidden for hitcircles? Essentially `StartTime - StartTime + 0`, I think I would prefer that expire was explicitly done down each relevant code path in any case, in-case the transformations change at some point.

Reply:
It works fine as I tested so far. It took me a while to understand how and why but it seems to be the `BeginDelayedSequence` that causes it to expire when all Transformations (and Delays) are completed. I didn't notice any cutoffs in normal play.